### PR TITLE
[WIP] RFC FS-1001 - String Interpolation

### PR DESCRIPTION
**Pull request by forki**
I "rebased" [the original prototype by @v2m](https://visualfsharp.codeplex.com/SourceControl/network/forks/vladima/primary/contribution/7063) on the current master and made it compile.
- F# Language RFC: https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1001-StringInterpolation.md
- [x] rebase original prototype
- [x] make it compile
- [x] create unit tests

image

### CODE REVIEW

**Comment by ashic** [comment]
Would it be preferable to use interpolation notation similar to other languages? Common things are $var, ${var}, #var , #{var} , possibly prefixing interpolated strings with a leader character (e.g. s"foo ${bar}") in order to avoid accidental interpolation. %(foo) seems like a new notation (I could be wrong...maybe it exists somewhere).

**Comment by forki** [comment]
@ashic I think C# syntax wasn't ready when @v2m built this. So yes making this more like C# string interpolation might be a good idea.
I just wanted to start by rebasing and bringing this back to life. Now we should discuss if we want this and how it should behave.

**Comment by forki** [comment]
A swift sample:

    let multiplier = 3
    let message = "\(multiplier) times 2.5 is \(Double(multiplier) * 2.5)"
    // message is "3 times 2.5 is 7.5"

**Comment by isaacabraham** [comment]
I'm split on the ```$"Hello {name}"``` C# usage - on the one hand, it's quite nice and goes well with the ```Console.WriteLine``` style syntax. On other other hand, the ```%``` style goes well with F#'s ```printfn``` style syntax. Maybe something like ```"Hello %{name}"``` would be reasonable (curly brace instead of brackets).

**Comment by enricosada** [comment]
I dont want to split the discussion or break momentum (thx @forki to resurrect this feature) but maybe we can discuss the implementation details in this pr and the high level design in fsharp/FSharpLangDesign#6 ?

**Comment by enricosada** [comment]
a question: are we supporting localization correctly?

[msdn about c# string interpolation](https://msdn.microsoft.com/en-us/library/dn961160.aspx)

It's not only about a better string format, need to support i18n too. C# does that because string interpolation does implicit type conversion to [FormattableString](https://msdn.microsoft.com/en-us/library/system.formattablestring.aspx)

```
var s = $"hello, {name}"
System.IFormattable s = $"Hello, {name}"
System.FormattableString s = $"Hello, {name}"
```

so

```
var product = "Raspberry Pi";
var price = 25.0;
Console.WriteLine(LocalizationHelper.Format(
    $"The price of {product} is {price}.";
));
```

> However, the string interpolation will be compiled to a FormattableString with “The price of {0} is {1:C}” as the format that will get passed to GetLocalized. In other words, the people doing the localization will still have to localize with numeric indices instead of the more readable named tokens

If the preferred way for f# is explitic, i think we need another function like `$` to formattablestring

**Comment by isaacabraham** [comment]
Yes sure. Frankly it's something that's easy to change anyway.

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/StringInterpolationTests.fs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    member val Y = s
+
+type B = {
+    X : string
+    Y : int
+}
+
+
+[<TestFixture>]
+type StringInterpolationTests() =
```
Comment (enricosada) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/StringInterpolationTests.fs]: I think these test should be in Compiler unit tests.
FSharp.Core shouldn't know about string interpolation.

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/StringInterpolationPropertyTests.fs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace FSharp.Core.Unittests.FSharp_Core.Microsoft_FSharp_Core
+
+open System
+open FSharp.Core.Unittests.LibraryTestFx
+open NUnit.Framework
+open FsCheck
+
+module PropertyTests =
```
Comment (enricosada) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/StringInterpolationPropertyTests.fs]: I think these test should be in Compiler unit tests.
FSharp.Core shouldn't know about string interpolation.

**Comment by forki** [comment]
just noticed the following

    type B = {
        X : string
        Y : int
    };;
    
    let b = { X = "world"; Y = 42 };;
    
    printf "hello %(b.X)";;  // OK
    
    printf "hello %(b.Y)";;  // crashes fsi

last line crashes with


    FatalExecutionEngineError occurred
    Message: Managed Debugging Assistant 'FatalExecutionEngineError' has detected a problem in 'D:\code\visualfsharp\src\fsharp\fsi\..\..\..\Debug\net40\bin\fsi.exe'.
    Additional information: Die Laufzeit hat einen schwerwiegenden Fehler entdeckt. Fehleradresse: "0x7319931f" in Thread "0x3ddc". Fehlercode: 0xc0000005. Bei diesem Fehler könnte es sich um ein Problem in der CLR oder in den unsicheren oder nicht verifizierbaren Teilen des Benutzercodes handeln. Übliche Ursachen dieses Bugs sind Marshallerfehler für COM-Interop oder PInvoke, die den Stapel beschädigen können.

seems we can't format ints.

    let x = 1;;

    printfn "%(x)" // crashes with NRE in printf.fs

**Comment by forki** [comment]
OK "fixed" that NRE exception by making it the same error class like `printf "hello %(b.Y)";;`

**Comment by forki** [comment]
@dsyme, @v2m @7sharp9: I assume the issue is at https://github.com/Microsoft/visualfsharp/pull/921/files#diff-5b9ab9dd9d7133aaf23add1048742031R6376  correct? what can we do?

**Comment by 7sharp9** [comment]
Erm ... dunno, I not looked at the WIP on this lately :-)

**Comment by KevinRansom** [comment]
I think a spec for this would be preferable to a prototype.  It was possible to discus the general approach that was taken for C# before a line of code was written.  At the very least decide if we want interpolation in string.Format, printf or both.  Should we mimic the C# syntax and format specs, or use the F# syntax ... or both, give some examples of what strings using interpolation look like to illustrate arguments for and against each spec requirement.

Let's discuss the language needs for the feature before we pile in and write a ton of code.  This seems like a pretty scary feature to "just evolve and debate"

Kevin

**Comment by forki** [comment]
As already written above: language design is at https://github.com/fsharp/FSharpLangDesign/blob/master/FSharp-4.1-Proposals/StringInterpolation.md 

I just rebase an existing explorative implementation on the latest master and added couple of tests to see how far it went.

**Comment by isaacabraham** [comment]
I don't think Steffen expects this to go straight through into master - I don't see a problem with anyone prototyping a sample string interpolation if for no other reason than to get us all thinking about how this could work in the real world (if at all).

**Comment by 7sharp9** [comment]
@forki Nice work bringing this feature back to life.  

I think this has been discussed thoroughly in the links at the top of the page, I was really disappointed that it was dropped in F#4 when it was so close to completion.

**Comment by vasily-kirichenko** [comment]
I think it should be _string_ interpolation (as it's in C#), not `printf` interpolation.

**Comment by forki** [comment]
@vasily-kirichenko yes that would make sense as well and unify sprintf and String.Format - since it would be already interpolated at that time.

**Comment by jarlestabell** [comment]
I assume most people here know the Scala solution to this, but just in case I want to point that one out, as I think it is quite elegant.
I assume it is drastically much heavier to implement though and it may overlap a bit with type provider territory.

The key point is that it has a concise syntax and is extensible.
So the same concept can be used both for string interpolation without formatting:

    val name = "James"
    println(s"Hello, $name")  // Hello, James

with formatting:

    val height = 1.9d
    val name = "James"
    println(f"$name%s is $height%2.2f meters tall")  // James is 1.90 meters tall

The above examples are taken from:
http://docs.scala-lang.org/overviews/core/string-interpolation.html


And plain SQL queries: (http://slick.typesafe.com/doc/3.1.1/sql.html#string-interpolation)

    def insert(c: Coffee): DBIO[Int] =
     sqlu"insert into coffees values (${c.name}, ${c.supID}, ${c.price}, ${c.sales}, ${c.total})"

and even things like ReactJS-like JSX/TSX xml-based syntax:

    xml”””
	      <body>
    		<a href = “some link”> ${linktext} </a>
	    </body>
	    ”””

(example from http://docs.scala-lang.org/sips/completed/string-interpolation.html)

**Comment by enricosada** [comment]
@jarlestabell scala syntax **and extensibility** it's really interesting, `StringContext` it's like [System.FormattableString](https://msdn.microsoft.com/en-us/library/system.formattablestring(v=vs.110).aspx) and more like `sprintf`.
The extensibility is really nice to have.

@jarlestabell can you post your comment in the string interpolation discussion in f# lang design repo? fsharp/FSharpLangDesign#6 so this pr can remain focused on implementation.

**Comment by jarlestabell** [comment]
@enricosada Done, thanks! :)

**Comment by jon49** [comment]
Another option for adding formatting:

```
type Person = {name: string; age: int} 
let person = {name = "Robert"; age=5} 
printfn "Hi %(person, A), your age is %(person.age, 10i) with spaces or age %(person.age) without." 
printfn : unit -> string
```

It would be nice to keep the formatting options. Whether this way or the ways that were suggested above.

Can't wait to get this feature.

**Comment by KevinRansom** [comment]
@forki,
@dsyme 
@vladima 

I'm closing this due to inactivity, if you get time to do further work on it, please reactivate.

Thanks

Kevin

**Comment by dsyme** [comment]
@KevinRansom Let's keep open any WIP PRs that correspond to [an approved F# RFC](https://github.com/fsharp/FSharpLangDesign/tree/master/RFCs).  Inactivity is OK for RFCs - they can take months or years to get through the system.

**Comment by AviAvni** [comment]
@forki @dsyme Can I help with this?

**Comment by forki** [comment]
I'm not sure if @dsyme has decided which syntax we will use? Is this
settled in RFC?

**Comment by vasily-kirichenko** [comment]
I think we should just finally pick one of the syntaxes and release it. Finally.

**Comment by dsyme** [comment]
@vasily-kirichenko @forki @AviAvni The RFC is here: https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1001-StringInterpolation.md

The basic proposed syntax is covered there.    However the RFC really needs considerable work. I'd particularly recommend that someone go over all the comments in this thread, and in the RFC discussion thread, and the original codeplex thread, and capture all the remaining issues and list them at the end of the RFC.  We can then resolve them one by one.

**Comment by vasily-kirichenko** [comment]
The implementation works with string expressions only because of https://github.com/forki/visualfsharp/commit/6542316ea089d6051be59a53b0b6866aa2be6448#diff-5b9ab9dd9d7133aaf23add1048742031R6583 where coercion from `string` to `obj` is performed:
```fsharp
mkCoerceIfNeeded cenv.g cenv.g.obj_ty cenv.g.string_ty s
```
What we need here is real type of each expression `s`. I have no idea how to get it, it seems there are no types at this stage at all.

**Comment by KevinRansom** [comment]
Closing

**Comment by isaacabraham** [comment]
@kevinransom again - why has this been closed?

**Comment by dsyme** [comment]
@isaacabraham  Same response as here https://github.com/Microsoft/visualfsharp/pull/908#issuecomment-259445419.  The issue is tracked by an RFC which is definitely still active, and we can resubmit this work (or a  variation of it) for F# 4.2

**Comment by realvictorprm** [comment]
I'm going to rebase this on master again and just implement for test purposes the C# syntax. There's nothing particular wrong with extending this later again.

**Comment by realvictorprm** [comment]
@forki do you want to give access to your PR or should I open another separate PR?

**Comment by forki** [comment]
it's closed. go ahead and submit your own

**Comment by realvictorprm** [comment]
Alright then, thank Sir!

**Pull request closed without merging**